---
title: "dNdSCV on LUAD"
author: "Sehyun Oh"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

GitHub page for dNdScr: https://github.com/im3sanger/dndscv
dNdScr vignette: http://htmlpreview.github.io/?http://github.com/im3sanger/dndscv/blob/master/vignettes/dNdScv.html

```{r echo=FALSE, results="hide", warning=FALSE}
suppressPackageStartupMessages({
    library(dndscv)
    library(dplyr)
})
```

# Set-up environment
Outputs are in two directories: `tumor_only` (`i=1`) and `matching_normal` (`i=2`)

```{r setup_1}
# Select runtype
i = 1

runtype = c("tumor_only", "matching_normal")[i]
wd = "~/data2/PureCN_manuscript/Revision/dNdSCV"
dnds_out_dir = file.path("/nobackup/16tb_b/CNVworkflow_revision/dNdSCV/outputs/LUAD", runtype)
```

# Build a reference for dNdSCV
```{r eval=FALSE}
# source("~/data2/PureCN_manuscript/Revision/dNdSCV/script/buildReference.R")
```


# Prepare SNV mutations table from PureCN
`buildSNVtable.R` script requires a runtype (`i`), output directory (`out_dir`), and
PureCN output data directory (`data_dir`). Ouput from this script is a list of SNV 
mutation tables (`snv_list`). 

- input: `i`, `out_dir`, `data_dir`   
- output: `snv_list`   

```{r buildSNVTable}
out_dir = "~/data2/PureCN_manuscript/Revision/dNdSCV"
data_dir = file.path("/nobackup/16tb_a/CNVworkflow_LUAD/purecn_output/PureCN", runtype)

source("~/data2/PureCN_manuscript/Revision/dNdSCV/script/buildSNVTable_LUAD.R")
```


# Format mutation table
## Load a list of SNV tables
```{r setup_2}
if (runtype == "matching_normal") {
    snv_list = readRDS(file.path(wd, "data", paste0("LUAD_mutTable_", runtype, ".rds")))
} else if (runtype == "tumor_only") {
    snv_list = readRDS(file.path(out_dir, "data", paste0("LUAD_mutTable_", runtype, "_rescalePriors.rds")))
}

names(snv_list) = stringr::str_extract(names(snv_list), "TCGA.{11}")
```

## PoN filtered
Checking if some of the false positives are found in the pool of normals (--> potentially
sequencing artifacts)
```{r Markus_script, eval=FALSE}
source("/home/riestma1/pon.R")
```

```{r pon.R_script, echo=FALSE, eval=FALSE}
library(VariantAnnotation)
x <- readRDS("LUAD_mutTable_tumor_only_rescalePriors.rds")  # list of data frame containing SNVs from each sample
xx <- GRanges(do.call(rbind, x))   # combine all the information in x into GRanges object
y <- read.csv("pathToPoN.csv", as.is=T)[,2]

# takes a while
# vcfs <- lapply(y, function(f) subsetByOverlaps(readVcf(f), xx))
# load pre-computed:
vcfs <- readRDS("/home/riestma1/LUAD_vcfs.rds")   # list of GRanges containing overlapping site between each sample and their SNP calls

# count overlaps for all variants with all VCFs
oa <- rowSums(do.call(cbind, lapply(vcfs, function(f) overlapsAny(xx, f))))   # the number of samples stil in the SNV calls??
xx$pon.count <- oa

# CD8B seems to match: 108 in tumor
subsetByOverlaps(xx, GRanges("chr2:86858179"))
length(subsetByOverlaps(xx, GRanges("chr2:86858179")))

saveRDS(xx, "LUAD_pon.rds")
```


```{r PoN_filtered}
xx = readRDS("/home/riestma1/LUAD_pon.rds")
x = xx[which(xx$pon.count < 5)]   # remove SNVs seen in >=5 in PoN

snv_list_gr = split(x, x$Sampleid)   # split each sample into GRangesList
snv_list = lapply(snv_list_gr, function(x) {data.frame(Sampleid = x$Sampleid,
                                                       chr = seqnames(x),
                                                       start = start(x),
                                                       end = end(x),
                                                       REF = x$REF,
                                                       ALT = x$ALT)})
```

## Format a list of SNV tables
* Input: `snv_list`, which is a SNPs/MNPs table before formatting      
* Output: `res`, which is a properly formatted SNPs/MNPs table, and potentially 
`mannual_snv_check`, which contains the samples with fewer than 3 SNVs. You need 
to check it manually (there will be a message in this case) and add back to the `res`.

```{r eval=TRUE}
source("~/data2/PureCN_manuscript/Revision/dNdSCV/script/formatSNVTable.R")
```

```{r eval=TRUE}
if (exists("mannual_snv_check")) {
    for (i in seq_along(mannual_snv_check)) {
        samplename = names(mannual_snv_check)[i]
        snv_list[[samplename]] = mannual_snv_check[[i]]
    }
}
snv_list = res   # assign 
```


# Run dNdSCV
* Input: `snv_list`, which is the properly formatted SNPs/MNPs table   
* Output: `dndsout`, which is an output from `dndscv::dndscv` function
* **Excluded samples** : We have run dNdScv with default parameters. This includes
removing ultra-hypermutator samples and subsampling mutations when encountering too
many mutations per gene in the same sample. These were designed to protect against 
loss of sensitivity from ultra-hypermutators and from clustered artefacts in the 
input mutation table, but there are occasions when the user will prefer to relax these.

```{r eval=TRUE}
source("~/data2/PureCN_manuscript/Revision/dNdSCV/script/rundNdSCV.R")
```

```{r echo=FALSE, eval=FALSE}
saveRDS(dndsout, file = file.path(dnds_out_dir, paste0("dndsout_LUAD_", runtype, "_rescalePriors_PoNFiltered.rds")))
```
