---
title: "dNdSCV on LUAD"
author: "Sehyun Oh"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

GitHub page for dNdScr: https://github.com/im3sanger/dndscv

```{r echo=FALSE, results="hide", warning=FALSE}
suppressPackageStartupMessages({
    library(dndscv)
    library(dplyr)
})
```

# Set-up environment
Outputs are in two directories: `tumor_only` (`i=1`) and `matching_normal` (`i=2`)

```{r setup_1}
# Select runtype
i = 1

runtype = c("tumor_only", "matching_normal")[i]
wd = "~/data2/PureCN_manuscript/Revision/dNdSCV"
dnds_out_dir = file.path("/nobackup/16tb_b/CNVworkflow_revision/dNdSCV/outputs/LUAD", runtype)
```



# Build a reference for dNdSCV
```{r eval=FALSE}
path_cds_table = file.path(wd, "mart_export.txt")
path_genome_fasta = "~/reference/GRCh38_no_alt/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna"

# Build GRCh38 reference
buildref(cdsfile = path_cds_table, 
         genomefile = path_genome_fasta, 
         outfile = file.path(wd, "GRCh38_refcds.rda"), 
         excludechrs = "MT")
```





# Prepare SNV mutations table from PureCN
`buildSNVtable.R` script requires a runtype (`i`), output directory (`out_dir`), and
PureCN output data directory (`data_dir`). Ouput from this script is a list of SNV 
mutation tables (`snv_list`). 

- input: `i`, `out_dir`, `data_dir`   
- output: `snv_list`   

```{r eval=FALSE}
out_dir = "~/data2/PureCN_manuscript/Revision/dNdSCV"
data_dir = file.path("/nobackup/16tb_a/CNVworkflow_LUAD/purecn_output/PureCN", runtype)

source("~/data2/PureCN_manuscript/Revision/dNdSCV/script/buildSNVTable.R")
```


# Format mutation table
## Load a list of SNV tables
```{r setup_2}
snv_list = readRDS(file.path(wd, paste0("LUAD_mutTable_", runtype, ".rds")))
names(snv_list) = stringr::str_extract(names(snv_list), "TCGA.{11}")
```

## Format a list of SNV tables
* Input: `snv_list`, which is a SNPs/MNPs table before formatting      
* Output: `res`, which is a properly formatted SNPs/MNPs table

```{r eval=TRUE}
source("~/data2/PureCN_manuscript/Revision/dNdSCV/script/formatSNVTable.R")
snv_list = res   # assign 
```


# Run dNdSCV
* Input: `snv_list`, which is the properly formatted SNPs/MNPs table   
* Output: `dndsout`, which is an output from `dndscv::dndscv` function

```{r eval=TRUE}
source("~/data2/PureCN_manuscript/Revision/dNdSCV/script/rundNdSCV.R")
```

```{r echo=FALSE, eval=FALSE}
saveRDS(dndsout, file = file.path(dnds_out_dir, paste0("dndsout_LUAD_", runtype, ".rds")))
```


## Excluded samples 
We have run dNdScv with default parameters. This includes removing ultra-hypermutator
samples and subsampling mutations when encountering too many mutations per gene in the
same sample. These were designed to protect against loss of sensitivity from 
ultra-hypermutators and from clustered artefacts in the input mutation table, but 
there are occasions when the user will prefer to relax these.   



# EDA on dNdSCV outputs
`sel_cv` is gene-wise selection results using dNdScv.

```{r}
sel_cv = dndsout$sel_cv
signif_genes = sel_cv[sel_cv$qglobal_cv < 0.1, c("gene_name", "qglobal_cv")]
rownames(signif_genes) = NULL

print(signif_genes)
write.csv(signif_genes, file = file.path(dnds_out_dir, paste0("signif_genes_LUAD_", runtype, ".csv")))
```






